drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -1442,787 +1571,1118 @@ void RasterizerCanvasGLES2::canvas_render_items(Item *p_item_list, int p_z, cons
+
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::USE_LIGHTING, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::USE_SHADOWS, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_NEAREST, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF3, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF5, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF7, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF9, false);
+		//			state.canvas_shader.set_conditional(CanvasShaderGLES2::SHADOW_FILTER_PCF13, false);
+
+		//			state.canvas_shader.bind();
+
+		//			ris.last_blend_mode = -1;
+		//		}
+	}
+
+	if (reclip) {
+		join = false;
+	}
+
+	// non rects will break the batching anyway, we don't want to record item changes, detect this
+	if (_detect_batch_break(ci)) {
+		join = false;
+		r_batch_break = true;
+	}
+
+	return join;
+}
+
+bool RasterizerCanvasGLES2::_detect_batch_break(Item *ci) {
+	int command_count = ci->commands.size();
+
+	// It is hard to know what this number should be, empirically,
+	// and this has not been fully investigated. It works to join single sprite items when set to 1 or above.
+	// Note that there is a cost to increasing this because it has to look in advance through
+	// the commands.
+	// On the other hand joining items where possible will usually be better up to a certain
+	// number where the cost of software transform is higher than separate drawcalls with hardware
+	// transform.
+
+	// if there are more than this number of commands in the item, we
+	// don't allow joining (separate state changes, and hardware transform)
+	// This is set to quite a conservative (low) number until investigated properly.
+	const int MAX_JOIN_ITEM_COMMANDS = 16;
+
+	if (command_count > MAX_JOIN_ITEM_COMMANDS) {
+		return true;
+	} else {
+		Item::Command *const *commands = ci->commands.ptr();
+
+		// do as many commands as possible until the vertex buffer will be full up
+		for (int command_num = 0; command_num < command_count; command_num++) {
+
+			Item::Command *command = commands[command_num];
+			CRASH_COND(!command);
+
+			switch (command->type) {
+
+				default: {
+					return true;
 				} break;
-				case RasterizerStorageGLES2::Shader::CanvasItem::BLEND_MODE_PMALPHA: {
-					glBlendEquation(GL_FUNC_ADD);
-					if (storage->frame.current_rt && storage->frame.current_rt->flags[RasterizerStorage::RENDER_TARGET_TRANSPARENT]) {
-						glBlendFuncSeparate(GL_ONE, GL_ONE_MINUS_SRC_ALPHA, GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
-					} else {
-						glBlendFuncSeparate(GL_ONE, GL_ONE_MINUS_SRC_ALPHA, GL_ZERO, GL_ONE);
-					}
+				case Item::Command::TYPE_RECT: {
 				} break;
-			}
+			} // switch
+
+		} // for through commands
+
+	} // else
+
+	return false;
+}
+
+void RasterizerCanvasGLES2::_canvas_render_item(Item *ci, RIState &ris) {
```

Comment:
It's concerning that this logic is now copied three times (when attempting to join, when rendering batched and here). I wonder if there is any way we could combine. 

On the other hand, I do appreciate how simple it makes the implementation.

Reply:
This section here is only repeated as a temporary measure for backward compatibility (i.e. turning batching off and finding regressions with the old renderer).

Once it has been fully tested and we are sure it works for everyone, there should be no need to retain the legacy codepath, and it could potentially be better to just remove it. As you say it is problematic trying to maintain 2 duplicate codepaths.

Reply:
Im okay with commenting this saying that it needs to be removed before being merged into 3.2. Its fine for regression testing

Reply:
Ok I'll comment that. :+1: